src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/ServerSentEventsMessageFormatter.cs

```diff
@@ -12,113 +12,101 @@ namespace Microsoft.AspNetCore.Http.Connections.Internal
+            while (source.TryGet(ref position, out ReadOnlyMemory<byte> memory))
             {
-                return;
+                // Last segment
+                if (offset + memory.Length >= totalLength)
+                {
+                    return memory;
+                }
+
+                offset += memory.Length;
             }
 
-            // We can't just use while(payload.Length > 0) because we need to write a blank final "data: " line
-            // if the payload ends in a newline. For example, consider the following payload:
-            //   "Hello\n"
-            // It needs to be written as:
-            //   data: Hello\r\n
-            //   data: \r\n
-            //   \r\n
-            // Since we slice past the newline when we find it, after writing "Hello" in the previous example, we'll
-            // end up with an empty payload buffer, BUT we need to write it as an empty 'data:' line, so we need
-            // to use a condition that ensure the only time we stop writing is when we write the slice after the final
-            // newline.
+            throw new InvalidOperationException("Could not get last segment from sequence.");
+        }
+
+        private static async Task WriteMessageToMemory(Stream output, ReadOnlySequence<byte> payload)
+        {
             var keepWriting = true;
             while (keepWriting)
             {
-                var span = payload.Span;
-                // Seek to the end of buffer or newline
-                var sliceEnd = span.IndexOf(LineFeed);
-                var nextSliceStart = sliceEnd + 1;
-                if (sliceEnd < 0)
-                {
-                    sliceEnd = payload.Length;
-                    nextSliceStart = sliceEnd + 1;
+                var sliceEnd = payload.PositionOf(LineFeed) ?? payload.End;
 
-                    // This is the last span
-                    keepWriting = false;
-                }
-                if (sliceEnd > 0 && span[sliceEnd - 1] == '\r')
+                ReadOnlySequence<byte> lineSegment;
+                if (sliceEnd == payload.End)
                 {
-                    sliceEnd--;
-                }
-
-                var slice = payload.Slice(0, sliceEnd);
-
-                if (nextSliceStart >= payload.Length)
-                {
-                    payload = ReadOnlyMemory<byte>.Empty;
+                    lineSegment = payload;
+                    payload = ReadOnlySequence<byte>.Empty;
+                    keepWriting = false;
                 }
                 else
                 {
-                    payload = payload.Slice(nextSliceStart);
+                    lineSegment = payload.Slice(payload.Start, sliceEnd);
+
+                    if (sliceEnd != payload.Start)
+                    {
+                        // Check if the line ended in \r\n. If it did then trim the \r
+                        var memory = GetLastSegment(in lineSegment, out var offset);
```

Comment:
Why are you specifying in here? It's implicit and doesn't need to be specified at the call site.

Reply:
Ah, VS squiggles it without the in but it still compiles. I'll remove.